### PR TITLE
Sumtree checkpointing

### LINKED ISSUE #11: Consider checkpointing HexSumTree every term
Currently the HexSumTree is a singleton-type object which we only modify in the `heartbeat()` function so the state of the tree stays constant during the term allowing deterministic juror drafts throughout the term. This comes with the following drawbacks:

- Because all modifications to the tree are batched for the `heartbeat()` of a future term, we need to limit how many mutations can be performed in a given term to avoid `heartbeat()` to require more gas than the block gas limit and bricking the court.

- Disputes need to draft their jurors strictly during the correct term because the juror sum tree can mutate in the transition to the next term. If the jurors of a dispute aren't drafted during the term, the dispute needs to be rescheduled for a future term.

By checkpointing the state of the tree every term we would be able to:

- Immediately perform tree modifications for the **next** term, no longer scheduling modifications to a future term's `heartbeat()`.

- Draft jurors for a dispute at any term in the future, as the drafting function would be able to perform sortition on a past state of the tree.

- Being able to prove to a contract how many tokens a juror had active and for how long, which would be really useful for distributing passive rewards (as discussed in #9)

Downsides:

- Gas cost:

  - I am particularly worried about the added `SLOAD` cost while traversing the tree performing a sortition, as for every node in the tree we'd need to get its checkpointed value (even though it is a binary search, it would increase the number `SLOAD`s by at least a factor of 2 and grows logarithmically with the number of balance updates of a juror has had).

  - The `SSTORE` cost for modifying tree entries wouldn't be that different for trees of a reasonable size specially taking into account that the Court would no longer have to manage queues in storage. The `SSTORE` overhead would be an extra `~20k * log16(N)` (`N` being the number of jurors) in the worst case (the first update for the next term), which would add up to an extra ~100k gas for a tree of ~1.04M (`16 ^ 5`) jurors. A positive side-effect is that this gas would be paid by the juror directly in almost all instances.

- Removing the requirement to draft an adjudication round during an specific term, presents some challenges of its own with the current design:
  - Deferred juror unstaking: because at the time drafting can only happen during one specific term and adjudication rounds that fail to draft on time need to be rescheduled, the moment a juror is deactivated we know exactly how many tokens they have at stake (and we can allow them to withdraw all their tokens that aren't at risk of being slashed). If an adjudication round no longer needs to be drafted during an specific term, we'd need to prevent the juror from unstaking for at least the duration of a full adjudication round (commit + reveal + appeal) which is a reasonable time span for cancelling/rescheduling the round if the draft didn't happen.

  - Higher risk of missing block hashes: the block hash used as a randomness seed for a term needs to be queried and saved within 257 blocks after the term is transitioned. If a round no longer needs to be drafted within the term, we may need to add an extra incentive to perform a secondary `hearbeat()` to store the randomness seed for all terms with adjudication rounds.

- Scheduling: at the moment, jurors can schedule their activation and deactivation to a custom term. This is a nice side-effect of not being able to perform activations immediately, as we don't care whether a modification happens in the next term because it needs to be scheduled anyway. If we were to add checkpointing, we'd likely remove the ability to schedule activations and deactivations.

### PR DESCRIPTION
**Pull request by bingen**
Builds on top of #25 
Fixes #11

### CODE REVIEW

**Comment by bingen** [comment]
Following up from [here](https://github.com/aragon/aragon-court/issues/11#issuecomment-477800316), first optimization measure implemented: grouping checkpoints by term, instead of by block number. This is certainly needed, as only one historical value per term is needed, potentially reducing the size by a factor of up to ~256, but not enough.
Replying last scenario from here, with only one block difference between updates, we see a substantial gain:
```
initial block number 0, term 0

Tree depth:    6
Tree next key: 1,048,586
Tree total sum:  360
Inserts
Size:    10

|         |   Total | Function |
|---------|---------|---------|
| Min     | 125,381 | 102,322 |
| Max     | 189,896 | 166,837 |
| Average | 151,317 | 128,258 |

Sets
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     |  86,796 |  63,823 |
| Max     | 225,937 | 202,899 |
| Average |  87,913 |  64,883 |

Sortitions
Size:    31

|         |   Total | Function |
|---------|---------|---------|
| Min     | 191,514 | 168,259 |
| Max     | 301,104 | 277,785 |
| Average | 209,190 | 185,924 |

final block number 300, term 1

```
But if we increase the time difference between updates to 243 blocks, we similar performance. The following example would mean that 10 different jurors have updated their balances (either "manually" themselves or by means of automatic fee redistribution) 30 times each in different terms (almost always). Actually, a total of 284 different terms have had an update on the tree. If every single term had activity (which sounds like a little bit too much), that would be ~12 days. If for instance an average of 5 terms a day registered updates, that scenario would be reached in ~57 days (i.e., less than 2 months).
```
initial block number 0, term 0

Tree depth:    6
Tree next key: 1,048,586
Tree total sum:  360
Inserts
Size:    10

|         |   Total | Function |
|---------|---------|---------|
| Min     | 125,381 | 102,322 |
| Max     | 189,896 | 166,837 |
| Average | 151,317 | 128,258 |

Sets
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     |  86,796 |  63,823 |
| Max     | 225,937 | 202,899 |
| Average | 219,043 | 196,011 |

Sortitions
Size:    287

|         |   Total | Function |
|---------|---------|---------|
| Min     | 191,514 | 168,259 |
| Max     | 1,025,806 | 1,002,487 |
| Average | 874,056 | 850,731 |

final block number 72900, term 284

```

**Comment by bingen** [comment]
# Weighted binary search on Checkpointing

Following up from the last example (multiple terms), next attempt was to use a weighted binary search for Checkpointing, see commit e136702
The results were slightly better for this example, although it's not clear that it would be generally beneficial. The improvement relies on the distribution of the timestamps being more or less uniform.


Sortitions
Size:    287

|         |   Total | Function |
|---------|---------|---------|
| Min     | 190,448 | 167,193 |
| Max     | 900,364 | 877,045 |
| Average | 676,792 | 653,467 |

final block number 72900, term 284

**Comment by bingen** [comment]
# Capped history length

Another idea was to see what happens if we restrict searches to a max number of hops back in the Checkpoint array, in the example, 30, which would mean in the worst case (having changes on every term) a little bit more than a day back.

## Over original Checkpointing
See commit 16ca6ed.

Sortitions
Size:    30

|         |   Total | Function |
|---------|---------|---------|
| Min     | 311,722 | 288,117 |
| Max     | 898,136 | 874,591 |
| Average | 761,343 | 737,741 |

final block number 72900, term 284

## Over Checkpointing with weighted binary search
See commit 480a27f.

Sortitions
Size:    30

|         |   Total | Function |
|---------|---------|---------|
| Min     | 478,033 | 454,424 |
| Max     | 794,136 | 770,591 |
| Average | 692,269 | 668,667 |

final block number 72900, term 284

**Comment by bingen** [comment]
# 4 children tree

Having Sum Tree of 4 children instead of 16 surprisingly didn't seem to improve the results for Sortition (nodes and updates were reduced to avoid timeouts):

initial block number 0, term 0
Tree depth:    11
Tree next key: 1,048,584
Tree total sum:  200
Inserts
Size:    8

|         |   Total | Function |
|---------|---------|---------|
| Min     | 168,816 | 145,735 |
| Max     | 401,027 | 377,946 |
| Average | 226,390 | 203,309 |

Sets
Size:    160

|         |   Total | Function |
|---------|---------|---------|
| Min     | 130,209 | 107,236 |
| Max     | 368,691 | 345,654 |
| Average | 355,148 | 332,119 |

Sortitions
Size:    152

|         |   Total | Function |
|---------|---------|---------|
| Min     | 284,426 | 261,171 |
| Max     | 1,147,471 | 1,124,152 |
| Average | 1,049,244 | 1,025,926 |

final block number 38880, term 151

**Comment by bingen** [comment]
# Checkpoint using mappings

Another attempt, using mappings for checkpoints and compressing terms of size `uint32` (if terms are 1h long, it's more than enough), packed into an array of `uint256`'s. 

Writing can become more expensive, but sortition is quite better. We could explore combining with the other improvements above, like the weighted binary search.

initial block number 256, term 1
Tree depth:    6
Tree next key: 1,048,586
Tree total sum:  400
Inserts
Size:    10

|         |   Total | Function |
|---------|---------|---------|
| Min     | 162,494 | 139,457 |
| Max     | 468,008 | 444,971 |
| Average | 193,045 | 170,008 |

Sets
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     | 106,529 |  83,578 |
| Max     | 370,059 | 347,044 |
| Average | 253,288 | 230,280 |

Sortitions
Size:    287

|         |   Total | Function |
|---------|---------|---------|
| Min     | 130,734 | 107,440 |
| Max     | 688,298 | 665,004 |
| Average | 499,187 | 475,887 |

final block number 73156, term 285

**Comment by bingen** [comment]
As so far the results are not good enough, but we need Checkpointing for the final phase of the appealing process, we are most likely going to keep it but still enforce that drafting is done during the term (see [this comment](https://github.com/aragon/aragon-court/issues/11#issuecomment-476216096)). This way, as we are getting the last checkpoint, the sortition during regular drafts will still be cheap, as we don't need to do the binary search.

**Comment by bingen** [comment]
# Only one element per sortition

All previous benchmarks were drafting 10 elements on each sortition. If we run the same test pulling only one element (and without any improvement), these are the results:

initial block number 0, term 0
Tree depth:    6
Tree next key: 1,048,586
Tree total sum:  400
Inserts
Size:    10

|         |   Total | Function |
|---------|---------|---------|
| Min     | 125,403 | 102,322 |
| Max     | 326,637 | 303,556 |
| Average | 145,526 | 122,445 |

Sets
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     |  86,796 |  63,823 |
| Max     | 225,937 | 202,899 |
| Average | 219,043 | 196,011 |

Sortitions
Size:    287

|         |   Total | Function |
|---------|---------|---------|
| Min     |  41,908 |  18,653 |
| Max     | 140,846 | 117,527 |
| Average | 109,089 |  85,765 |

final block number 72900, term 284

The behaviour seems to be quite linear.

**Comment by bingen** [comment]
I we are sticking with [sortitions on last term](https://github.com/aragon/aragon-court/pull/28#issuecomment-481681473), last optimization is not worth, it's actually worse. Some numbers:

# Multiple sortition on last term only:

<table>
<tr><th>Regular checkpointing </th><th>Checkpointing with mappings</th></tr>
<tr><td>
initial block number 0, term 0

Tree depth:    6
Tree next key: 1,048,586
Tree total sum:  400
Inserts
Size:    10

|         |   Total | Function |
|---------|---------|---------|
| Min     | 125,403 | 102,322 |
| Max     | 326,637 | 303,556 |
| Average | 145,526 | 122,445 |

Sets
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     |  86,796 |  63,823 |
| Max     | 225,937 | 202,899 |
| Average | 225,467 | 202,435 |

Sortitions
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     | 238,145 | 215,008 |
| Max     | 333,871 | 310,734 |
| Average | 288,110 | 264,973 |

final block number 76800, term 300

</td><td>
initial block number 256, term 1

Tree depth:    6
Tree next key: 1,048,586
Tree total sum:  400
Inserts
Size:    10

|         |   Total | Function |
|---------|---------|---------|
| Min     | 162,494 | 139,457 |
| Max     | 468,008 | 444,971 |
| Average | 193,045 | 170,008 |

Sets
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     | 106,529 |  83,578 |
| Max     | 370,059 | 347,044 |
| Average | 260,710 | 237,702 |

Sortitions
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     | 288,786 | 265,693 |
| Max     | 527,340 | 504,247 |
| Average | 400,364 | 377,271 |

final block number 77056, term 301

</td></tr> </table>

File: contracts/lib/Checkpointing.sol [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        if (length == 0) {
+            return 0;
+        }
+
+        uint256 lastIndex = length - 1;
+
+        // short-circuit
+        if (time >= self.history[lastIndex].time) {
+            return self.history[lastIndex].value;
+        }
```
[the remark this answers is not in the document]
Comment (izqui) [reply contracts/lib/Checkpointing.sol]: As we spoke offline, it makes more sense that we do this directly in `getLast()`

File: contracts/lib/Checkpointing.sol [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+        add192(self, time, uint192(value));
+    }
+
+    function get(History storage self, uint64 time) internal view returns (uint256) {
+        return uint256(get192(self, time));
+    }
+
+    function getLast(History storage self) internal view returns (uint256) {
+        if (self.history.length > 0) {
```
Comment (izqui) [inline contracts/lib/Checkpointing.sol]: For functions like these that are going to be called many times during sortition, we should make sure that we don't do any extra storage reads.

File: contracts/Court.sol [hunk 3]
```diff
@@ -1023,7 +1023,7 @@ contract Court is ERC900, ApproveAndCallFallBack {
 
             if (update.delta > 0) {
                 // account.sumTreeId always > 0: as only a juror that activates can get in this queue (and gets its sumTreeId)
-                sumTree.update(account.sumTreeId, update.delta, update.positive);
+                sumTree.update(account.sumTreeId, term, update.delta, update.positive);
```
[the remark this answers is not in the document]
Comment (bingen) [reply contracts/Court.sol]: To be honest, I didn't spend too much time about this, as with checkpointing we would remove queues so I just wanted it to compile to be able to measure gas.

[the remark this answers is not in the document]
Comment (bingen) [reply contracts/Court.sol]: I already added this suggestion, although it seems that when this function is called `prevTermId` is just `term`, right?

**Comment by bingen** [comment]
# Mulit-sortition

Another attempt of optimizing this, trying to go down the tree only once, so kind of doing the sortition in parallel, in order to "reuse" checkpointing binary searches.
There's a little trick here, because in this "fake" big tree all significant nodes are adjacent, which makes this optimization work way better than it would in a regular situation this big, but I think we can somehow force that (making sure that we keep "fairness", in the sense that all nodes keep the same probability). I think it definitely opens up some room for exploration.

## Getting all past element

initial block number 0, term 0

Tree depth:    6
Tree next key: 1,048,586
Tree total sum:  400
Inserts
Size:    10

|         |   Total | Function |
|---------|---------|---------|
| Min     | 125,491 | 102,322 |
| Max     | 326,725 | 303,556 |
| Average | 145,614 | 122,445 |

Sets
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     |  86,840 |  63,823 |
| Max     | 225,981 | 202,899 |
| Average | 219,087 | 196,011 |

Sortitions
Size:    287

|         |   Total | Function |
|---------|---------|---------|
| Min     | 117,829 |  93,843 |
| Max     | 204,862 | 180,940 |
| Average | 191,600 | 167,673 |

final block number 72900, term 284

## Sortition on last term only

It's also a significant improvement when only getting last checkpoint. Here as Checkpointing has a shortcut, binary searches won't happen, but it's still better:

 initial block number 0, term 0

Tree depth:    6
Tree next key: 1,048,586
Tree total sum:  400
Inserts
Size:    10

|         |   Total | Function |
|---------|---------|---------|
| Min     | 125,513 | 102,322 |
| Max     | 326,747 | 303,556 |
| Average | 145,636 | 122,445 |

Sets
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     |  86,818 |  63,823 |
| Max     | 225,959 | 202,899 |
| Average | 225,489 | 202,435 |

Sortitions
Size:    300

|         |   Total | Function |
|---------|---------|---------|
| Min     | 104,345 |  80,385 |
| Max     | 124,997 | 101,037 |
| Average | 117,541 |  93,581 |

final block number 76800, term 300

## TODO
- [ ] Some cleaning
- [ ] Try not to use `variables` array, which is ugly and more expensive. The reason for it is the `Stack too deep` error.
- [ ] Explore the possibility of combining it with some of the previous attempts.
- [ ] Add specific tests for multi-sortition, checking that all values in returned array are correct. So far only regular tests (with an array of 1 element) were run piping them to this new function, plus some manual tests.
- [ ] Try to build a better more realistic tree.

**Comment by facuspagnuolo** [comment]
Closing this PR since `development` branch was already updated with these commits manually

**Pull request closed without merging**